xcape/core.py

```diff
@@ -129,40 +129,78 @@ def calc_cape(p, t, td, ps, ts, tds, **kwargs):
+        print('too many optional arguments')
+        
+    elif kwargs['vertical_lev']!=True:
+        print('please select type of vertical grid, pressure level, or model level')
+        
+    elif (len(args)==0)&(kwargs['vertical_lev'] == 'pressure'):
+        print('vertical grid is set to pressure levels,\n'+
+                +'but location in fortran values (1: nlev) of where p <= ps is missing')
+        
+    elif (len(args)==1)&(args[0].shape!=ps.shape)&(kwargs['vertical_lev'] == 'pressure'):
+        print('vertical grid is set to pressure levels,\n'+
+                +'but location in fortran values (1: nlev) of where p <= ps is not the correct shape\n'+
+                +'(pres_lev_pos.shape should be equal to ps.shape)')
+        
+        
     if _any_dask_array(p, t, td, ps, ts, tds):
-        return _calc_cape_gufunc(p, t, td, ps, ts, tds, **kwargs)
+        return _calc_cape_gufunc(p, t, td, ps, ts, tds, *args, **kwargs)
     else:
-        return _calc_cape_numpy(p, t, td, ps, ts, tds, **kwargs)
+        return _calc_cape_numpy(p, t, td, ps, ts, tds, *args, **kwargs)
     
 
-def _calc_cape_gufunc(p, t, td, ps, ts, tds, **kwargs):
-    signature = "(i),(i),(i),(),(),()->(),()"
-    output_dtypes = ('f4','f4')
-    if kwargs['source']=='most-unstable':
-        signature += ",(),()"
-        output_dtypes = output_dtypes + ('i4','f4')
-        
-    return da.apply_gufunc(_calc_cape_numpy, signature,
-                           p, t, td, ps, ts, tds,
-                           output_dtypes=('f4','f4'),
-                           axis=-1,
-                           vectorize=False,
-                           **kwargs)
-
+def _calc_cape_gufunc(p, t, td, ps, ts, tds, *args, **kwargs):
+    
+    if (kwargs['vertical_lev']=='sigma'):
+        signature = "(i),(i),(i),(),(),()->(),()"
+        output_dtypes = ('f4','f4')
+        if kwargs['source']=='most-unstable':
+            signature += ",(),()"
+            output_dtypes = output_dtypes + ('i4','f4')
+
+        return da.apply_gufunc(_calc_cape_numpy, signature,
+                               p, t, td, ps, ts, tds, 
+                               output_dtypes=('f4','f4'),
+                               axis=-1,
+                               vectorize=False,
+                               **kwargs)
+    
+    elif (kwargs['vertical_lev']=='pressure'):
+        print('here')
+        pres_lev_pos = args[0]
+        signature = "(i),(i),(i),(),(),(),()->(),()"
+        output_dtypes = ('f4','f4')
+        if kwargs['source']=='most-unstable':
+            signature += ",(),()"
+            output_dtypes = output_dtypes + ('i4','f4')
+
+        return da.apply_gufunc(_calc_cape_numpy, signature,
+                               p, t, td, ps, ts, tds, pres_lev_pos,
+                               output_dtypes=('f4','f4'),
+                               axis=-1,
+                               vectorize=False,
+                               **kwargs)
```

Comment:
this is the signature and call for the case in which vertical_lev == pressure. 
I added another () for input, and the argument pres_lev_pos in the call.

Reply:
I recommend not putting the whole function in a giant `if` block.

Just modify the signature and other args as needed. Then just have a single `return da.apply_gufunc` with the properly customized arguments.

Reply:
in one it has only p,t,td, in the other has pres_lev_pos

Reply:
so I really think I cannot move the function out of the if loop.

Reply:
You can if you use `*args`

Reply:
ok I put *args.  I left the call of calc_cape as is, but changed to *args in the other routines. To have all *args is a bit confusing to me right now.